doc/feature/synapses.md

```diff
@@ -0,0 +1,215 @@
+Synapse support in Brain
+============
+
+The low-level data access for synapses provided by Brion needs a complementary
+API in Brain to support various common higher-level uses cases, like semantic
+synapse object access or full-circuit synapse density voxelization, to name the
+two extremes.
+
+## Requirements
+
+* Semantic synapse objects that give read-access to all properties
+* Array access which allows vectorization
+* Filtering of synapses for pathway use case (projected synapses of two targets)
+* Lazy-loading to support out-of-core scenarios
+* User-transparent KVS caching
+
+## New dependencies
+
+None
+
+## API
+
+```cpp
+
+namespace brain
+{
+    // < postGID, afferent_index >
+    typedef std::pair< uint32_t, size > SynapseGID;
+
+    enum SynapsesPrefetch
+    {
+        SP_ATTRIBUTES = 1 << 0,
+        SP_POSITIONS = 1 << 1,
+        SP_ALL = SP_ATTRIBUTES | SP_POSITIONS
+    }
+}
+
+class Circuit
+{
+    SynapsesStream getAfferentSynapses( const GIDSet& gids,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+    SynapsesStream getEfferentSynapses( const GIDSet& gids,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+    SynapsesStream getProjectedSynapses( const GIDSet& preGIDs,
+                                         const GIDSet& postGIDs,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+};
+
+// read-only, moveable proxy object of data storage in Synapses container
+class Synapse
+{
+    uint32_t preGID() const;
+    uint32_t preSectionID() const;
+    uint32_t preSegmentID() const;
+    float preDistance() const;
+    Vector3f preSurfacePosition() const;
+    Vector3f preCenterPosition() const;
+
+    uint32_t postGID() const;
+    uint32_t postSectionID() const;
+    uint32_t postSegmentID() const;
+    float postDistance() const;
+    Vector3f postSurfacePosition() const;
+    Vector3f postCenterPosition() const;
+
+    SynapseGID gid() const;
+    float conductance() const;
+    float utilization() const;
+    float depression() const;
+    float facilitation() const;
+    float decay() const;
+    int efficacy() const;
+};
+
+// read-only, moveable, non-copyable
+class Synapses
```

Comment:
Just a side note, I think it's implied by being const, but this class must be thread-safe internally in my opinion. Otherwise it's unusable with parallel algorithms.